core/lib/Cardano/Multisig.hs

```diff
@@ -0,0 +1,128 @@
+data Script =
+      RequireSignatureOf VerificationKeyHash
+    | RequireAllOf [Script]
+    | RequireAnyOf [Script]
+    | RequireMOf Word8 [Script]
+    deriving stock (Generic, Show, Eq)
+
+instance NFData Script
+
+-- | A 'VerificationKeyHash' type represents verification key hash
+-- that participate in building MultisigScript. The hash is expected to have size of
+-- 28-byte.
+--
+-- @since 3.0.0
+newtype VerificationKeyHash = VerificationKeyHash ByteString
+    deriving (Generic, Show, Eq)
+
+instance NFData VerificationKeyHash
+
+fromVerificationKey :: Shelley 'MultisigK XPub -> VerificationKeyHash
+fromVerificationKey = VerificationKeyHash . blake2b224
+
+verificationKeyHashFromBytes :: ByteString -> VerificationKeyHash
+verificationKeyHashFromBytes = VerificationKeyHash . invariantSize hashSize
+
+toCBOR' :: Script -> CBOR.Encoding
+toCBOR' (RequireSignatureOf (VerificationKeyHash verKeyHash)) =
+    encodeMultiscriptCtr 0 2 <> CBOR.encodeBytes verKeyHash
+toCBOR' (RequireAllOf contents) =
+    encodeMultiscriptCtr 1 2 <> encodeFoldable toCBOR' contents
+toCBOR' (RequireAnyOf contents) =
+    encodeMultiscriptCtr 2 2 <> encodeFoldable toCBOR' contents
+toCBOR' (RequireMOf m contents) =
+    encodeMultiscriptCtr 3 3 <> CBOR.encodeInt (fromInteger $ toInteger m)
+    <> encodeFoldable toCBOR' contents
+
+-- | This function realizes what cardano-node's `Api.serialiseToCBOR script` realizes
+-- This is basically doing the symbolically following:
+-- toCBOR [0,multisigScript]
+toCBOR :: Script -> ByteString
+toCBOR script =
+    CBOR.toStrictByteString $ encodeMultisigBeginning <> toCBOR' script
+
+encodeMultisigBeginning :: CBOR.Encoding
+encodeMultisigBeginning = CBOR.encodeListLen 2 <> CBOR.encodeWord8 0
+
+encodeMultiscriptCtr :: Word -> Word -> CBOR.Encoding
+encodeMultiscriptCtr ctrIndex listLen =
+    CBOR.encodeListLen listLen <> CBOR.encodeWord ctrIndex
+
+encodeFoldable :: (Foldable f) => (a -> CBOR.Encoding) -> f a -> CBOR.Encoding
+encodeFoldable encode xs = wrapArray len contents
+  where
+    (len, contents) = foldl' go (0, mempty) xs
+    go (!l, !enc) next = (l + 1, enc <> encode next)
+
+    wrapArray :: Word -> CBOR.Encoding -> CBOR.Encoding
+    wrapArray len' contents' =
+        if len' <= 23
+        then CBOR.encodeListLen len' <> contents'
+        else CBOR.encodeListLenIndef <> contents' <> CBOR.encodeBreak
+
+-- | This function realizes what cardano-node's
+-- `Api.serialiseToRawBytes $ Api.scriptHash script` realizes
+-- This is basically doing the symbolically the following (using symbols from toCBOR):
+-- digest $ (nativeMultisigTag <> toCBOR multisigScript )
+toScriptHash :: Script -> ByteString
+toScriptHash script = digest $ nativeMultiSigTag <> toCBOR'' script
+  where
+      nativeMultiSigTag :: ByteString
+      nativeMultiSigTag = "\00"
+
+      digest = BA.convert . hash @_ @Blake2b_224
+
+      toCBOR'' :: Script -> ByteString
+      toCBOR'' script' = CBOR.toStrictByteString $ toCBOR' script'
```

Comment:
I'm a little bit confused about the difference between this one and toCBOR above.

The `encodeMultisigBeginning` seems similar but slightly different to `nativeMultiSigTag`.

Reply:
so I wanted two-step golden test here, namely for (Api stands for cardano-api):
1. Api.serialiseToCBOR script
2. Api.serialiseToRawBytes $ Api.scriptHash script
But there is problem here how cbor is handled. Symbolically, 
`Api.serialiseToCBOR script` ->   `toCBOR [0,multisigScript]`
`Api.scriptHash script` -> `digest $ (nativeMultisigTag <> toCBOR multisigScript )`
and this is reflected in the `toCBOR'` and `toCBOR''`
This is how cardano-node handles it, spent some time debugging that, and can only adjust to it